planner/core/rule_aggregation_elimination.go

```diff
@@ -155,3 +272,174 @@ func (a *aggregationEliminator) optimize(p LogicalPlan) (LogicalPlan, error) {
+	case inner.Name == ast.AggFuncBitOr && outer.Name == ast.AggFuncBitOr:
+	case inner.Name == ast.AggFuncBitXor && outer.Name == ast.AggFuncBitXor:
+	case inner.Name == ast.AggFuncGroupConcat && outer.Name == ast.AggFuncGroupConcat:
+		if inner.HasDistinct {
+			return nil
+		}
+	case inner.Name == ast.AggFuncSum && outer.Name == ast.AggFuncSum:
+		if inner.HasDistinct || outer.HasDistinct {
+			return nil
+		}
+	case inner.Name == ast.AggFuncCount && outer.Name == ast.AggFuncSum:
+		// TODO: add cast to support `sum(count(_)) -> count(_)`
+		return nil
+	default:
+		return nil
+	}
+	return combined
+}
+
+// genColMaps generates exprMap(`col -> definition expr`) and aggMap(`col -> definition aggFunc`) from column definitions.
+func genColMaps(ptn *nestedAggPattern) (map[string]expression.Expression, map[string]*aggregation.AggFuncDesc) {
+	exprMap := make(map[string]expression.Expression, len(ptn.proj.Schema().Columns))
+	for idx, col := range ptn.proj.Schema().Columns {
+		exprMap[string(col.HashCode(nil))] = ptn.proj.Exprs[idx]
+	}
+	aggMap := make(map[string]*aggregation.AggFuncDesc, len(ptn.inner.Schema().Columns))
+	for idx, col := range ptn.inner.Schema().Columns {
+		aggMap[string(col.HashCode(nil))] = ptn.inner.AggFuncs[idx]
+	}
+
+	return exprMap, aggMap
+}
+
+// exprSubstitute tries to substitutes expressions occurred in given `exprs` with values in `exprMap`.
+// If the substitution happened, it returns true and the substituted expressions, otherwise it returns false and original exprs.
+func exprSubstitute(ctx sessionctx.Context, exprMap map[string]expression.Expression, exprs []expression.Expression) (bool, []expression.Expression) {
+	var newExprs []expression.Expression
+	replaced := false
+	for idx, expr := range exprs {
+		if e, ok := exprMap[string(expr.HashCode(ctx.GetSessionVars().StmtCtx))]; ok {
+			replaced = true
+			if newExprs == nil {
+				newExprs = make([]expression.Expression, len(exprs))
+				copy(newExprs, exprs)
+			}
+			newExprs[idx] = e
+			continue
+		}
+		sf, ok := expr.(*expression.ScalarFunction)
+		if !ok {
+			continue
+		}
+		if subReplaced, subExprs := exprSubstitute(ctx, exprMap, sf.GetArgs()); subReplaced {
+			replaced = true
+			newSf := expression.NewFunctionInternal(ctx, sf.FuncName.L, sf.GetType(), subExprs...)
+			if newExprs == nil {
+				newExprs = make([]expression.Expression, len(exprs))
+				copy(newExprs, exprs)
+			}
+			newExprs[idx] = newSf
+		}
+	}
+	if replaced {
+		return true, newExprs
+	}
+	return false, exprs
+}
+
+// aggSubstitute tries to substitutes expressions occurred in given `exprs` with the expression argument of aggregated functions defined in `aggMap`,
+// parameter `predicate` can be used to check each of the aggregated function.
+// This function returns:
+//  * true / substituted exprs if the substitution happens.
+//  * false / original exprs if substitution doesn't happen.
+//  * false / nil if predicate is not nil and it failed.
+func aggSubstitute(
```

Comment:
This function shares a lot of logic with `exprSubstitute`, is it possible to merge them together? or using `ColumnSubstitute` instead?

Reply:
Done, thanks